### ISSUE TITLE
expose more helper functions for composing sql fragments

### ISSUE THREAD

**Issue by purge** [jawj/zapatos#4]
i like the fact this library encourages composability and its a big advantage over ORMs. Taking some of these patterns out of shortcuts and exposing them would be nice - such as the multiple insert form `values (1,2,3), (4,5,6)`

**Comment by jawj** [comment]
Thanks for all this feedback. I see that we're practically neighbours!

Your particular example here seems like a nice idea, but on the other hand:

- Is there a specific context in which you'd want to use this that the `insert` shortcut doesn't cover?

- What other patterns would you suggest for this treatment?

- I'm a bit wary of helper functions proliferating so much that it becomes hard to keep track of them (and then also hard to follow custom SQL queries) ...

**Comment by purge** [comment]
ha yeah i noticed that, i'd offer to get you a beer in other circumstances considering how many headaches writing the types for this must've caused. In this case I was stuck because upsert doesn't have a 'do nothing' option (i'll raise that as a separate issue) but i'm sure there will be other use-cases. i'll keep this up to date as i discover them.

**Comment by jawj** [comment]
I had a bit more of a think about your specific request for a helper that generates `(1, 2, 3), (4, 5, 6)` (etc.) for manual `INSERT` queries.

You can in fact make your own, like this:

```typescript
import * as db from './zapatos/src';
import * as s from './zapatos/schema';
import * as zu from './zapatos/src/utils';

const multiVals = (insertables: s.Insertable[]) =>
      zu.mapWithSeparator(
        zu.completeKeysWithDefault(insertables),
        db.sql`, `,
        v => db.sql`(${db.vals(v)})`,
      );
```

BUT — you then also need something similar for the columns listing (the `("a", "b", "c")` of `INSERT INTO ("a", "b", "c") ...`), again using `completeKeysWithDefault`, otherwise things go awry when your object keys don't match across every `Insertable` in the input array.

Once you've dealt with this too, you've more-or-less recreated the whole `insert` shortcut. Which means I'm not sure this helper is really all that much help!

So I'm going to close this issue for now.

Happy for you to reopen with other helper suggestions, and also for you to open that 'do nothing' option request for `upsert` you mentioned (which I can see a need for, though it's not then strictly an `upsert`).
